user/vcs.man
.\" $Id: vcs.man,v 11.6 2025/09/28 22:53:51 tom Exp $
.TH VCS 1 2025-09-28 "" "User commands"
.SH NAME
\fBvcs\fR \-
version control system utility
.SH SYNOPSIS
\fBvcs\fI [options] [file-specifications]\fR
.SH DESCRIPTION
\fBVcs\fR is a simple, easy to use interface to \fIrcs\fR
(\fBr\fRevision \fBc\fRontrol \fBs\fRystem).
.PP
\fBVcs\fR is a source-control tool which can be used by either
the configuration manager or developers.
Developers should be able to unlock files that they have locked
(i.e., with \fBcheckin\fR), or to add new subdirectories to a
configuration, or to delete them.
.PP
\fBVcs\fR uses the \fIrcs\fR utility to maintain
a special file in the "RCS" directory (named "RCS,v").
This file controls permission on the creation or deletion of
files -- or of the entire directory.
.SH OPTIONS
.TP 4n
.B \-d
delete archive-directory leaf.
If you are on the access list of the vcs-file within which the directory
is appended, and if no files have been archived in the RCS directories
within, you may delete the given directories.
.TP
.B \-i
insert archive-directory leaf.
If you are on the access list of the vcs-file within which the directory
is appended, \fBvcs\fR will create new archive directories, each of
which has an RCS subdirectory.
.TP
.B \-n
no-op mode
shows the operations that would be performed, but does not attempt them.
.TP
.B \-q
quiet mode
suppresses messages.
.TP
.B \-u
unlock specified files.
You must be either the owner of the file, or of at least one lock.
If the former, all locks are removed.
.TP
.B \-x
direct \fBvcs\fR
to retain the pathname when translating between working and archive
files, e.g, "\fBpath/name\fR" becomes "\fBpath/RCS/name,v\fR".
If the "\-\fBx\fR" option is not set, "\fBpath/name\fR"
becomes "\fB./RCS/name,v\fR", and "\fBpath/RCS/name,v\fR"
becomes "\fB./name\fR".
.PP
You must specify exactly one of the
options "-\fBd\fR", "-\fBi\fR" or "-\fBu\fR".
.SH ENVIRONMENT
\fBVcs\fR is written in C, and runs on POSIX systems.
.PP
Environment variables imported by \fBvcs\fR include:
.TP
.B RCS_DEBUG
set this variable (e.g., to 1 or 2) to activate debugging-traces in the tool.
.TP
.B RCS_DIR
identifies the name of the RCS subdirectory, and indirectly the name of
the \fBvcs\fR file within that directory.
.SH FILES
\fBVcs\fR uses the following files
.TP
.B permit
to initialize a new vcs-file (i.e., so that its baseline version is compatible).
.TP
.B rcs
to copy access lists from a parent directory to a newly-inserted directory.
.SH ANTICIPATED CHANGES
None.
.SH AUTHORS
Thomas E. Dickey <dickey@invisible-island.net>
.SH SEE ALSO
checkin, checkout, ci\ (1), co\ (1), rcs\ (1)


CHANGES
2025-09-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* package/debian/copyright: formatting

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	check-manpage

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	fixup-th-man

	* user/rcsget.man, user/vcs.man: check-manpage

	* user/rcsget.man, user/vcs.man: fixup-th-man

	* configure: regen

	* aclocal.m4, configure.in: use CF_PROG_INSTALL

2025-07-09  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: update to 2025-07-10

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: fix license

2025-01-17  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* configure.in: use CF_WITH_SYSTYPE early

2025-01-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/src/checkin.c, src/checkout/src/checkout.c,
	  src/checkup/src/checkup.c, src/checkout/src/rcsget.c,
	  src/copy/src/copy.c, src/link2rcs/src/link2rcs.c,
	  src/permit/src/permit.c, src/vcs/src/delete.c, src/vcs/src/unlock.c,
	  src/checkin/src/rcsput.c, src/baseline/src/baseline.c:
	gcc15-warnings

2024-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2024/07/27

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2024/05/27

2024-04-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023-12-27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023-12-02

2023-01-22  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec: bump

	* configure: regen

	* bin/makefile.in:
	build-fix for Solaris: rename variable $(USER) to $(AS_USER)

	* package/debian/copyright, COPYING: bump

	* aclocal.m4, configure.in: use pkg-config to find td_lib

	* package/debian/changelog: bump

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/compat: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* aclocal.m4: build-fixes for NetBSD

	* configure: regen

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/makefile.in, src/vcs/src/makefile.in,
	  src/baseline/test/makefile.in, src/baseline/makefile.in,
	  src/checkin/test/makefile.in, src/checkout/src/makefile.in,
	  src/checkout/test/makefile.in, src/checkup/src/makefile.in,
	  src/checkup/test/makefile.in, src/copy/src/makefile.in,
	  src/copy/test/makefile.in, src/link2rcs/src/makefile.in,
	  src/link2rcs/test/makefile.in, support/makefile.in:
	use VPATH

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/makefile.in, src/checkout/makefile.in,
	  src/checkup/makefile.in, src/copy/makefile.in,
	  src/link2rcs/makefile.in, src/permit/makefile.in, src/vcs/makefile.in:
	use VPATH

	* package/debian/compat: bump

	* src/baseline/src/makefile.in, src/checkin/src/makefile.in,
	  src/permit/src/makefile.in, src/permit/test/makefile.in,
	  src/vcs/test/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/checkup.man: fix typo

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/rcsput.man, user/rcsget.man, user/permit.man, user/link2rcs.man,
	  user/checkup.man, user/checkin.man, user/copy.man, user/checkout.man,
	  user/baseline.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

2019-12-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* src/checkin/src/rcsput.c, src/checkin/src/checkin.c:
	use DYN-argv lists to replace catarg()

	* src/vcs/src/vcs.c, src/permit/src/permit.c:
	use show_argv2(), to handle verb vs argv[0]

	* src/checkout/src/checkout.c, src/checkout/src/rcsget.c:
	use DYN-argv lists to replace catarg()


src/vcs/src/vcs.c
/*
 * Title:	vcs.c (version-control-system utility)
 * Author:	T.E.Dickey
 * Created:	06 Sep 1989
 * Modified:
 *		05 Dec 2019, use "show_argv2()"
 *		04 Dec 2019, use "executev()"
 *		14 Dec 2014, coverity warnings
 *		22 Sep 1993, gcc warnings
 *		17 Oct 1991, split out 'vcs_insert.c' and 'vcs_unlock.c'
 *		10 Oct 1991, began rewrite (never did finish this!)
 *
 * Function:
 */

#define	MAIN
#include <vcs.h>

MODULE_ID("$Id: vcs.c,v 11.9 2019/12/05 10:14:13 tom Exp $")

/************************************************************************
 *	utility procedures						*
 ************************************************************************/

void
set_command(void)
{
    RCS_argc = 1;
    if (q_opt) {
	add_params("-q");
    }
}

void
set_option(int option, const char *value)
{
    char temp[BUFSIZ];
    FORMAT(temp, "-%c%s", option, value);
    add_params(temp);
}

void
add_params(const char *value)
{
    RCS_argv[RCS_argc++] = txtalloc(value);
    RCS_argv[RCS_argc] = NULL;
}

int
do_command(void)
{
    if (verbose || no_op) {
	show_argv2(stdout, RCS_verb, RCS_argv);
    }
    return no_op ? 0 : executev(RCS_argv);
}

/*
 * Invokes 'do_command' (which in turn runs the RCS_argv), setting
 * our uid/gid to the effective user (usually!).
 */
void
invoke_command(const char *verb, const char *path)
{
    RCS_verb = verb;
    RCS_argv[0] = txtalloc(path);
    if (for_admin2(do_command, RCS_uid, RCS_gid) < 0)
	failed(RCS_verb);
}

/******************************************************************************/
void
ShowPath(const char *command, const char *name)
{
    if (verbose) {
	char temp[BUFSIZ];
	if (debug) {
	    if (getwd(temp))
		(void) pathcat(temp, temp, name);
	} else
	    (void) relpath(temp, original, name);
	PRINTF("%% %s %s\n", command, temp);
    }
}

/******************************************************************************/
/* test to see if an argument is a directory */
int
DirExists(const char *path)
{
    struct stat sb;
    return (stat(path, &sb) >= 0 && (sb.st_mode & S_IFMT) == S_IFDIR);
}

/******************************************************************************/
/* test to see if an argument is a directory (ignore it if so!) */
int
IsDirectory(const char *path)
{
    struct stat sb;
    if (stat(path, &sb) < 0 || (sb.st_mode & S_IFMT) != S_IFDIR)
	return FALSE;
    VERBOSE(".. is a directory\n");
    return TRUE;
}

/******************************************************************************/
/* find who owns the archive and its directory and govern our use of set-uid
 * rights accordingly */
int
Access(char *archive, int ok_if_noop)
{
    char *path;
    struct stat sb;

    ShowPath("access", archive);
    /* if we cannot even read the archive, give up */
    if (access(archive, R_OK) < 0 && !ok_if_noop) {
	failed(archive);
	/*NOTREACHED */
    }

    path = pathhead(archive, &sb);
    RCS_prot = (sb.st_mode & 0777);
    if (access(path, R_OK | W_OK | X_OK) < 0) {
	/* hope that effective user is this */
	RCS_uid = sb.st_uid;
	RCS_gid = sb.st_gid;
    } else {
	/* real user is sufficient here */
	RCS_uid = getuid();
	RCS_gid = getgid();
    }
    return TRUE;
}

/******************************************************************************/
time_t
DateOf(char *name)
{
    struct stat sb;
    if (stat(name, &sb) >= 0 && (sb.st_mode & S_IFMT) == S_IFREG)
	return sb.st_mtime;
    return 0;
}

/******************************************************************************/
static void
DoArg(char *name)
{
    char base[MAXPATHLEN];

    VERBOSE("** processing %s\n", name);

    if (strlen(name) < sizeof(base)) {
	switch (operation) {
	case Unlock:
	    if (!IsDirectory(name))
		UnLockFile(name);
	    break;
	case Delete:
	    DeleteDir(name);
	    break;
	case Insert:
	    if (InsertDir(name, base))
		VERBOSE(".. completed %s.x %s\n", base, name);
	    if (chdir(original) != 0)
		failed(original);
	    break;
	default:
	    break;
	}
    }
}

/******************************************************************************/
static void
usage(void)
{
    static const char *msg[] =
    {
	"Usage: vcs [options] [names]",
	"",
	"Options:",
	"  -d       delete archive-directory leaf",
	"  -i       insert archive-directory leaf",
	"  -n       no-op mode",
	"  -q       quiet mode",
	"  -u       unlock specified files",
	"  -x       assume archive and working file are in path2/RCS and path2",
	"           (normally assumes ./RCS and .)"
    };
    unsigned j;
    for (j = 0; j < sizeof(msg) / sizeof(msg[0]); j++)
	FPRINTF(stderr, "%s\n", msg[j]);
    exit(FAIL);
}

/******************************************************************************/
/*ARGSUSED*/
_MAIN
{
    register int c;

    debug = RCS_DEBUG;
    operation = Unknown;
    while ((c = getopt(argc, argv, "dinqux")) != EOF)
	switch (c) {
	case 'd':
	    operation = Delete;
	    break;
	case 'i':
	    operation = Insert;
	    break;
	case 'n':
	    no_op = TRUE;
	    break;
	case 'q':
	    q_opt = TRUE;
	    break;
	case 'u':
	    operation = Unlock;
	    break;
	case 'x':
	    x_opt = TRUE;
	    break;
	default:
	    usage();
	    /*NOTREACHED */
	}

    verbose = debug || !q_opt;
    if (operation == Unknown) {
	FPRINTF(stderr, "expected one of -u, -i, -d\n");
	usage();
    }

    if (!getwd(original))
	failed("getwd");

    if (optind < argc) {
	while (optind < argc)
	    DoArg(argv[optind++]);
    } else {
	FPRINTF(stderr, "expected %s name\n",
		(operation == Unlock) ? "file" : "directory");
	usage();
    }
    exit(SUCCESS);
    /*NOTREACHED */
}
